torch/distributions/wishart.py

```diff
@@ -0,0 +1,262 @@
+            self._event_shape[-1],
+            device=self._unbroadcasted_scale_tril.device,
+            dtype=self._unbroadcasted_scale_tril.dtype,
+        )
+        return torch.cholesky_solve(
+            identity, self._unbroadcasted_scale_tril
+        ).expand(self._batch_shape + self._event_shape)
+
+    @property
+    def mean(self):
+        return self.df.view(self._batch_shape + (1, 1,)) * self.covariance_matrix
+
+    @property
+    def variance(self):
+        V = self.covariance_matrix  # has shape (batch_shape x event_shape)
+        diag_V = V.diagonal(dim1=-2, dim2=-1)
+        return self.df.view(self._batch_shape + (1, 1,)) * (V.pow(2) + torch.einsum("...i,...j->...ij", diag_V, diag_V))
+
+    def _bartlett_sampling(self, sample_shape=torch.Size()):
+        p = self._event_shape[-1]  # has singleton shape
+
+        # Implemented Sampling using Bartlett decomposition
+        noise = self._dist_chi2.rsample(sample_shape).sqrt().diag_embed(dim1=-2, dim2=-1)
+        i, j = torch.tril_indices(p, p, offset=-1)
+        noise[..., i, j] = torch.randn(
+            torch.Size(sample_shape) + self._batch_shape + (p * (p - 1) // 2,),
+            dtype=noise.dtype,
+            device=noise.device,
+        )
+        chol = self._unbroadcasted_scale_tril @ noise
+        return chol @ chol.transpose(-2, -1)
+
+    def rsample(self, sample_shape=torch.Size(), max_try=10):
+        sample_shape = torch.Size(sample_shape)
+        sample = self._bartlett_sampling(sample_shape)
+
+        # Below part is to improve numerical stability temporally and should be removed in the future
+        is_singular = self.support.check(sample).logical_not()
+        if self._batch_shape:
+            is_singular = is_singular.amax(self._reduced_batch_dims)
+
+        if is_singular.any():
+            warnings.warn("Singular sample detected.")
+
+            for _ in range(max_try):
+                if sample_shape or not is_singular.all():
+                    new_sample = self._bartlett_sampling(is_singular[is_singular].shape)
+                    sample = sample[is_singular.logical_not()]
+
+                    is_singular = self.support.check(new_sample).logical_not()
+                    if self._batch_shape:
+                        is_singular = is_singular.amax(self._reduced_batch_dims)
+
+                    sample = torch.cat(
+                        (sample, new_sample[is_singular.logical_not()]),
+                        dim=-(len(self._batch_shape) + 3)
+                    )
+
+                else:
+                    new_sample = self._bartlett_sampling(sample_shape)
+                    sample = new_sample
+
+                    is_singular = self.support.check(new_sample).logical_not()
+                    if self._batch_shape:
+                        is_singular = is_singular.amax(self._reduced_batch_dims)
+
+                if not is_singular.any():
+                    break
+
+        return sample
```

Comment:
We should probably add a warning to the docstring that in some cases we may end up returning nan values if we are not able to sample valid samples within max_try. In those cases, the user should validate the samples and either fix the value of `df` or adjust `max_try` accordingly.

Reply:
You're right. I will add the description in docstring. I am looking for the example of warning and I will follow the practice of it if I can find one.

Reply:
As a reference, here's a warning in the cuda graphs [documentation]( https://pytorch.org/docs/stable/generated/torch.cuda.graph.html). You can add that using `.. warning::`. See [source](https://github.com/pytorch/pytorch/blob/e3210ca184c87bfa5ba16cf6a83ed83b21865dcf/torch/cuda/graphs.py#L22).

Reply:
Thanks for your help. I almost miss this part.